plugins/witness_trx_plugin/witness_trx_plugin.cpp

```diff
@@ -0,0 +1,243 @@
+   bool delete_previous = false;
+
+   chain_plugin& chainplug = app().get_plugin<eosio::chain_plugin>();
+
+   boost::filesystem::path trx_witness_path;
+
+   chain::time_point last_known_head_time;
+   chain::block_id_type lib_block_id;
+   std::list<std::vector<char>> sig_action_data_waiting_on_catch_up;
+
+   std::thread thread;
+   boost::asio::io_context ctx;
+   boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard{ctx.get_executor()};
+
+   struct outstanding_witness_transaction {
+      outstanding_witness_transaction(const chain::digest_type& trx_id, const chain::time_point& trx_expires, std::list<std::vector<char>>&& action_data) :
+        trx_id(trx_id), trx_expires(trx_expires), action_data(std::move(action_data)) {}
+      chain::digest_type trx_id;
+      chain::time_point trx_expires;
+      std::list<std::vector<char>> action_data;
+   };
+
+   struct by_trx_id;
+   struct by_trx_expiration;
+
+   typedef multi_index_container<
+      outstanding_witness_transaction,
+      indexed_by<
+         hashed_unique<tag<by_trx_id>, key<&outstanding_witness_transaction::trx_id>, std::hash<chain::transaction_id_type>>,
+         ordered_non_unique<tag<by_trx_expiration>, key<&outstanding_witness_transaction::trx_expires>>
+      >
+   > outstanding_witness_transaction_index_type;
+   outstanding_witness_transaction_index_type outstanding_witness_transaction_index;
+
+   bool caught_up() const {
+      return last_known_head_time >= fc::time_point::now() - fc::seconds(5);
+   }
+
+   void on_accepted_block(const chain::block_state_ptr& bsp) {
+      ctx.post([this, this_block_time=bsp->block->timestamp]() {
+         if(this_block_time > last_known_head_time)
+            last_known_head_time = this_block_time;
+
+         if(!sig_action_data_waiting_on_catch_up.empty() && caught_up()) {
+            while(!sig_action_data_waiting_on_catch_up.empty()) {
+               std::list<std::vector<char>> batch;
+               constexpr unsigned max_batch_size = 1000u;
+               auto it = sig_action_data_waiting_on_catch_up.begin();
+               for(unsigned int i = 0; i < max_batch_size && it != sig_action_data_waiting_on_catch_up.end(); ++i, ++it); //ick
+               batch.splice(batch.begin(), sig_action_data_waiting_on_catch_up, sig_action_data_waiting_on_catch_up.begin(), it);
+               submit_witness_trx(std::move(batch));
+            }
+         }
+
+         auto it = outstanding_witness_transaction_index.get<by_trx_expiration>().begin();
+         while(it != outstanding_witness_transaction_index.get<by_trx_expiration>().end() && it->trx_expires < chain::time_point::now()) {
+            wlog("Witness transaction not confirmed by producers; resubmitting");
+            outstanding_witness_transaction_index.get<by_trx_expiration>().modify(it, [this](witness_trx_plugin_impl::outstanding_witness_transaction& row) {
+               submit_witness_trx(std::move(row.action_data));
```

Comment:
Does it make sense to resubmit transactions when we aren't `caught_up`?  Conversely, what is the motivation for delaying new transactions when we aren't `caught_up`?

Reply:
Also if outgoing transactions get backlogged, would we want to batch them?